Stripe/STPPaymentContext.m

```diff
@@ -626,39 +632,27 @@ - (void)requestPayment {
-                            /**
-                             When createCardSources is false, the SDK:
-                             1. Sends the token to customers/[id]/sources. This
-                             adds token.card to the customer's sources list.
-                             Surprisingly, attaching token.card to the customer
-                             will fail.
-                             2. Returns token.card to didCreatePaymentResult,
-                             where the user tells their backend to create a charge.
-                             A charge request with the token ID and customer ID
-                             will fail because the token is not linked to the
-                             customer (the card is).
-                             */
-                            if ([source isKindOfClass:[STPToken class]]) {
-                                paymentResultSource = ((STPToken *)source).card;
-                            }
-                            STPPaymentResult *result = [[STPPaymentResult alloc] initWithSource:paymentResultSource];
-                            [self.delegate paymentContext:self didCreatePaymentResult:result completion:^(NSError * error) {
-                                // for Apple Pay, the didFinishWithStatus callback is fired later when Apple Pay VC finishes
-                                if (error) {
-                                    completion(error);
-                                } else {
-                                    completion(nil);
-                                }
-                            }];
+                [self handleApplePaySource:source
+                                completion:^(NSError * error) {
+                    // for Apple Pay, the didFinishWithStatus callback is fired later when Apple Pay VC finishes
+                    // The apple pay vc finishes when the user cancels and it finishes only once.
+                    // So if the user has already canceled, calling completion will not cause the apple pay vc to finish
+                    // Therefore, we must directly call our did finish with status method
+                    if (error) {
+                        completion(error);
+                        if (isCanceledAfterPaymentResultCreated) {
+                            [self didFinishWithStatus:STPPaymentStatusError error:error];
                         }
-                    });
+                    } else {
+                        completion(nil);
+                        // if isCanceledAfterPaymentResultCreated is true then we should call self.didFinishWithStatus immediately
+                        // because the user already canceled. Otherwise, self.didFinishWithStatus will be called later
+                        if (isCanceledAfterPaymentResultCreated) {
+                            [self didFinishWithStatus:STPPaymentStatusSuccess error:nil];
```

Comment:
How do you feel about DRYing up this method? I think it'll allow you to re-write and re-position the comments so it's easier to understand what's going on:

```objective-c
completion(error); // error might be nil

// ...
if (isCanceled...) {
  // ...
  [self didFinishWithStatus:(error != nil ? ...Error : ...Success) error:error];
}